File: ai_v2_0/regression_plot.py
#!/usr/bin/env python3
"""Fit a simple linear regression and save a scatter plot with regression line."""

import os
import sys
from pathlib import Path

cache_dir = Path(__file__).parent / ".matplotlib_cache"
cache_dir.mkdir(exist_ok=True)
os.environ.setdefault("MPLCONFIGDIR", str(cache_dir))

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


def usage() -> None:
    print("Usage: python regression_plot.py <filename> <x_column> <y_column>")


def main() -> None:
    if len(sys.argv) != 4:
        usage()
        sys.exit(1)

    filename = sys.argv[1]
    x_column = sys.argv[2]
    y_column = sys.argv[3]

    data_path = Path(filename)
    if not data_path.exists():
        print(f"Error: file not found: {filename}")
        sys.exit(1)

    data = pd.read_csv(data_path)
    if x_column not in data.columns or y_column not in data.columns:
        print(f"Error: columns must exist in {filename}")
        print(f"Available columns: {', '.join(data.columns)}")
        sys.exit(1)

    x = data[[x_column]].values
    y = data[y_column].values

    model = LinearRegression()
    model.fit(x, y)
    y_pred = model.predict(x)

    r2 = r2_score(y, y_pred)
    rmse = np.sqrt(mean_squared_error(y, y_pred))

    x_line = np.linspace(x.min(), x.max(), 100).reshape(-1, 1)
    y_line = model.predict(x_line)

    plt.figure(figsize=(8, 6))
    plt.scatter(data[x_column], data[y_column], alpha=0.7, edgecolors="k", linewidths=0.5)
    plt.plot(x_line, y_line, color="red", linewidth=2, label="Regression line")
    plt.xlabel(x_column)
    plt.ylabel(y_column)
    plt.title(f"Linear Regression: {y_column} vs {x_column}")
    plt.text(
        0.05,
        0.95,
        f"R² = {r2:.4f}\nRMSE = {rmse:.4f}",
        transform=plt.gca().transAxes,
        verticalalignment="top",
        bbox={"boxstyle": "round", "facecolor": "white", "alpha": 0.8},
    )
    plt.legend()
    plt.tight_layout()

    output_name = f"{data_path.stem}_regression_plot_python.png"
    plt.savefig(output_name, dpi=150)
    plt.close()

    print(f"Saved regression plot to {output_name}")
    print(f"R-squared: {r2:.4f}")
    print(f"RMSE: {rmse:.4f}")


if __name__ == "__main__":
    main()

File: ai_v2_0/regression_plot.R
#!/usr/bin/env Rscript
# Fit a simple linear regression and save a scatter plot with regression line.

args <- commandArgs(trailingOnly = TRUE)

if (length(args) != 3) {
  cat("Usage: Rscript regression_plot.R <filename> <x_column> <y_column>\n")
  quit(status = 1)
}

filename <- args[1]
x_column <- args[2]
y_column <- args[3]

if (!file.exists(filename)) {
  cat(sprintf("Error: file not found: %s\n", filename))
  quit(status = 1)
}

data <- read.csv(filename, stringsAsFactors = FALSE)

if (!(x_column %in% names(data)) || !(y_column %in% names(data))) {
  cat(sprintf("Error: columns must exist in %s\n", filename))
  cat(sprintf("Available columns: %s\n", paste(names(data), collapse = ", ")))
  quit(status = 1)
}

formula <- as.formula(paste(y_column, "~", x_column))
model <- lm(formula, data = data)

r2 <- summary(model)$r.squared
rmse <- sqrt(mean(residuals(model)^2))

x_vals <- data[[x_column]]
y_vals <- data[[y_column]]

output_name <- sprintf("%s_regression_plot_r.png", tools::file_path_sans_ext(basename(filename)))

png(output_name, width = 800, height = 600, res = 150)
plot(
  x_vals,
  y_vals,
  xlab = x_column,
  ylab = y_column,
  main = sprintf("Linear Regression: %s vs %s", y_column, x_column),
  pch = 19,
  col = adjustcolor("steelblue", alpha.f = 0.7)
)
abline(model, col = "red", lwd = 2)
legend(
  "topleft",
  legend = c("Observations", "Regression line"),
  col = c("steelblue", "red"),
  pch = c(19, NA),
  lty = c(NA, 1),
  lwd = c(NA, 2),
  bty = "n"
)
mtext(sprintf("R² = %.4f | RMSE = %.4f", r2, rmse), side = 3, line = 0.2, adj = 0)
dev.off()

cat(sprintf("Saved regression plot to %s\n", output_name))
cat(sprintf("R-squared: %.4f\n", r2))
cat(sprintf("RMSE: %.4f\n", rmse))
